prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -814,6 +814,83 @@ fi_addr_t rxr_pkt_insert_addr(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry
 	return rdm_addr;
 }
 
+static inline
+void rxr_pkt_handle_zero_copy_recv_completion(struct rxr_ep *ep,
+					      struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_base_hdr *base_hdr;
+	size_t expected_hdr_size, received_hdr_size;
+	/*
+	 * in zero copy receive mode, EFA device must be the
+	 * only source of incoming packets. So shm communication
+	 * cannot be used, therefore peer must NOT be local
+	 */
+	assert(!rxr_ep_get_peer(ep, pkt_entry->addr)->is_local);
+	ep->posted_bufs_efa--;
+
+	base_hdr = rxr_get_base_hdr(pkt_entry->pkt);
+	if (base_hdr->type != RXR_EAGER_MSGRTM_PKT) {
```

Comment:
A genernal question, why is it always `RXR_EAGER_MSGRTM_PKT` for zero copy receive mode? Is it because the sender can only send eager message in zero copy receive mode?

Reply:
Correct. Zero copy is used when app does not require ordered send (FI_ORDER_SAS) and does not require tag matching, and the maximum message size is less then mtu_size.

I need to check, but they cannot request RMA capability either.

In this mode, they can only send eager  msgrtm,

Reply:
I see, thanks

Reply:
Why would RMA capabilities be disabled?  That doesn't make sense.

Reply:
This is because RMA is emulated now by send/receive. When application call `fi_write` to a peer, a packet is sent to the peer. If the peer is in zero copy receive mode, user buffer will be used to receive that packet.

Reply:
ok, that makes sense.  Thanks.